File: apps/ws/src/Game.ts
import { Chess, Move, Square } from 'chess.js';
import {
  GAME_ENDED,
  INIT_GAME,
  MOVE,
} from './messages';
import { db } from './db';
import { randomUUID } from 'crypto';
import { SocketManager, User } from './SocketManager';

type GAME_STATUS = 'IN_PROGRESS' | 'COMPLETED' | 'ABANDONED' | 'TIME_UP';
type GAME_RESULT = "WHITE_WINS" | "BLACK_WINS" | "DRAW";

const GAME_TIME_MS = 10 * 60 * 60 * 1000;

export function isPromoting(chess: Chess, from: Square, to: Square) {
  if (!from) {
    return false;
  }

  const piece = chess.get(from);

  if (piece?.type !== 'p') {
    return false;
  }

  if (piece.color !== chess.turn()) {
    return false;
  }

  if (!['1', '8'].some((it) => to.endsWith(it))) {
    return false;
  }

  return chess
    .moves({ square: from, verbose: true })
    .map((it) => it.to)
    .includes(to);
}

export async function fetchPlayerRatings(userId: string, timeControl: string) {
  const rating = await db.rating.findFirst({
    where: {
      userId: userId
    }
  })
  if (rating) {
    if (timeControl == 'CLASSICAL') {
      return rating.classic
    } else if (timeControl == 'RAPID') {
      return rating.rapid
    } else if (timeControl == 'BLITZ') {
      return rating.blitz
    } else if (timeControl == 'BULLET') {
      return rating.bullet
    }
  }
}

export async function fetchExpectedRating(playerRating: number, opponenetRating: number) {
  let qA = Math.pow(10, (playerRating / 400))
  let qB = Math.pow(10, (opponenetRating / 400))
  let eA = qA / (qA + qB)
  let k = 20 // TODO:need to add a method to select different K-factor based on condition. 
  let win = Math.round(k * (1 - eA))
  let draw = Math.round(k * (0.5 - eA))
  let lose = Math.round(k * (0 - eA))
  return {
    win,
    draw,
    lose
  }
}

export async function updatePlayersRating(gameFromDb:any) {
  console.log("gameFromDb -- ",gameFromDb)
  let whitePlayerRating = await fetchPlayerRatings(gameFromDb?.whitePlayerId, gameFromDb?.timeControl) as number
  let blackPlayerRating = await fetchPlayerRatings(gameFromDb?.blackPlayerId, gameFromDb?.timeControl) as number

  const whitePlayerExpected = await fetchExpectedRating(whitePlayerRating, blackPlayerRating)
  const blackPlayerExpected = await fetchExpectedRating(blackPlayerRating, whitePlayerRating)


  if (gameFromDb.result == 'WHITE_WINS') {
    whitePlayerRating = whitePlayerRating + whitePlayerExpected.win
    blackPlayerRating = blackPlayerRating + blackPlayerExpected.lose
  } else if (gameFromDb.result == 'BLACK_WINS') {
    whitePlayerRating = whitePlayerRating + whitePlayerExpected.lose
    blackPlayerRating = blackPlayerRating + blackPlayerExpected.win
  } else {
    whitePlayerRating = whitePlayerRating + whitePlayerExpected.draw
    blackPlayerRating = blackPlayerRating + blackPlayerExpected.draw
  }

 let whitePlayerRatingDb = await db.rating.findFirst({
    where : {
      userId : gameFromDb?.whitePlayerId
    }
  })

  await db.rating.update({
    data:{
      classic : gameFromDb?.timeControl == 'CLASSICAL' ? whitePlayerRating : whitePlayerRatingDb?.classic,
      bullet : gameFromDb?.timeControl == 'BULLET' ? whitePlayerRating : whitePlayerRatingDb?.bullet,
      rapid : gameFromDb?.timeControl == 'RAPID' ? whitePlayerRating : whitePlayerRatingDb?.rapid,
      blitz : gameFromDb?.timeControl == 'BLITZ' ? whitePlayerRating : whitePlayerRatingDb?.blitz
    },
    where :{
      userId : gameFromDb.whitePlayerId
    }
  })

  let blackPlayerRatingDb = await db.rating.findFirst({
    where : {
      userId : gameFromDb?.whitePlayerId
    }
  })

  await db.rating.update({
    data:{
      classic : gameFromDb?.timeControl == 'CLASSICAL' ? blackPlayerRating : blackPlayerRatingDb?.classic,
      bullet : gameFromDb?.timeControl == 'BULLET' ? blackPlayerRating : blackPlayerRatingDb?.bullet,
      rapid : gameFromDb?.timeControl == 'RAPID' ? blackPlayerRating : blackPlayerRatingDb?.rapid,
      blitz : gameFromDb?.timeControl == 'BLITZ' ? blackPlayerRating : blackPlayerRatingDb?.blitz
    },
    where :{
      userId : gameFromDb?.blackPlayerId
    }
  })

}

export class Game {
  public gameId: string;
  public player1UserId: string;
  public player2UserId: string | null;
  public board: Chess;
  private moveCount = 0;
  private timer: NodeJS.Timeout | null = null;
  private moveTimer: NodeJS.Timeout | null = null;
  public result: GAME_RESULT | null = null;
  private player1TimeConsumed = 0;
  private player2TimeConsumed = 0;
  private startTime = new Date(Date.now());
  private lastMoveTime = new Date(Date.now());
  private timeControl = ''

  constructor(player1UserId: string, player2UserId: string | null, gameId?: string, startTime?: Date) {
    this.player1UserId = player1UserId;
    this.player2UserId = player2UserId;
    this.board = new Chess();
    this.gameId = gameId ?? randomUUID();
    if (startTime) {
      this.startTime = startTime;
      this.lastMoveTime = startTime;
    }
  }

  seedMoves(moves: {
    id: string;
    gameId: string;
    moveNumber: number;
    from: string;
    to: string;
    comments: string | null;
    timeTaken: number | null;
    createdAt: Date;
  }[]) {
    console.log(moves);
    moves.forEach((move) => {
      if (
        isPromoting(this.board, move.from as Square, move.to as Square)
      ) {
        this.board.move({
          from: move.from,
          to: move.to,
          promotion: 'q',
        });
      } else {
        this.board.move({
          from: move.from,
          to: move.to,
        });
      }
    });
    this.moveCount = moves.length;
    if (moves[moves.length - 1]) {
      this.lastMoveTime = moves[moves.length - 1].createdAt;
    }

    moves.map((move, index) => {
      if (move.timeTaken) {
        if (index % 2 === 0) {
          this.player1TimeConsumed += move.timeTaken;
        } else {
          this.player2TimeConsumed += move.timeTaken;
        }
      }
    });
    this.resetAbandonTimer();
    this.resetMoveTimer();
  }
  async updateSecondPlayer(player2UserId: string) {
    this.player2UserId = player2UserId;

    const users = await db.user.findMany({
      where: {
        id: {
          in: [this.player1UserId, this.player2UserId ?? ''],
        },
      },
    });

    try {
      await this.createGameInDb();
    } catch (e) {
      console.error(e);
      return;
    }

    const player1rating = await fetchPlayerRatings(this.player1UserId, this.timeControl) as number
    const player2rating = await fetchPlayerRatings(this.player2UserId, this.timeControl) as number

    SocketManager.getInstance().broadcast(
      this.gameId,
      JSON.stringify({
        type: INIT_GAME,
        payload: {
          gameId: this.gameId,
          whitePlayer: {
            name: users.find((user) => user.id === this.player1UserId)?.name,
            id: this.player1UserId,
            rating: player1rating,
            expectedRating: await fetchExpectedRating(player1rating, player2rating)
          },
          blackPlayer: {
            name: users.find((user) => user.id === this.player2UserId)?.name,
            id: this.player2UserId,
            rating: player2rating,
            expectedRating: await fetchExpectedRating(player2rating, player1rating)
          },
          fen: this.board.fen(),
          moves: [],
        },
      }),
    );
  }

  async createGameInDb() {
    this.startTime = new Date(Date.now());
    this.lastMoveTime = this.startTime;

    const game = await db.game.create({
      data: {
        id: this.gameId,
        timeControl: 'RAPID',
        status: 'IN_PROGRESS',
        startAt: this.startTime,
        currentFen: 'rnbqkbnr/pppppppp/8/8/8/8/PPPPPPPP/RNBQKBNR w KQkq - 0 1',
        whitePlayer: {
          connect: {
            id: this.player1UserId,
          },
        },
        blackPlayer: {
          connect: {
            id: this.player2UserId ?? '',
          },
        },
      },
      include: {
        whitePlayer: true,
        blackPlayer: true,
      },
    });
    this.gameId = game.id;
    this.timeControl = game.timeControl
  }

  async addMoveToDb(move: Move, moveTimestamp: Date) {
    
    await db.$transaction([
      db.move.create({
        data: {
          gameId: this.gameId,
          moveNumber: this.moveCount + 1,
          from: move.from,
          to: move.to,
          before: move.before,
          after: move.after,
          createdAt: moveTimestamp,
          timeTaken: moveTimestamp.getTime() - this.lastMoveTime.getTime(),
          san: move.san
        },
      }),
      db.game.update({
        data: {
          currentFen: move.after,
        },
        where: {
          id: this.gameId,
        },
      }),
    ]);
  }

  async makeMove(
    user: User,
    move: Move
  ) {
    
    // validate the type of move using zod
    if (this.board.turn() === 'w' && user.userId !== this.player1UserId) {
      return;
    }

    if (this.board.turn() === 'b' && user.userId !== this.player2UserId) {
      return;
    }

    if (this.result) {
      console.error(`User ${user.userId} is making a move post game completion`);
      return;
    }

    const moveTimestamp = new Date(Date.now());

    try {
      if (isPromoting(this.board, move.from, move.to)) {
        this.board.move({
          from: move.from,
          to: move.to,
          promotion: 'q',
        });
      } else {
        this.board.move({
          from: move.from,
          to: move.to,
        });
      }
    } catch (e) {
      console.error("Error while making move");
      return;
    }

    // flipped because move has already happened
    if (this.board.turn() === 'b') {
      this.player1TimeConsumed = this.player1TimeConsumed + (moveTimestamp.getTime() - this.lastMoveTime.getTime());
    }

    if (this.board.turn() === 'w') {
      this.player2TimeConsumed = this.player2TimeConsumed + (moveTimestamp.getTime() - this.lastMoveTime.getTime());
    }

    await this.addMoveToDb(move, moveTimestamp);
    this.resetAbandonTimer()
    this.resetMoveTimer();

    this.lastMoveTime = moveTimestamp;

    SocketManager.getInstance().broadcast(
      this.gameId,
      JSON.stringify({
        type: MOVE,
        payload: { move, player1TimeConsumed: this.player1TimeConsumed, player2TimeConsumed: this.player2TimeConsumed },
      }),
    );

    if (this.board.isGameOver()) {
      const result = this.board.isDraw()
        ? 'DRAW'
        : this.board.turn() === 'b'
          ? 'WHITE_WINS'
          : 'BLACK_WINS';

      this.endGame("COMPLETED", result);
    }

    this.moveCount++;
  }

  getPlayer1TimeConsumed() {
    if (this.board.turn() === 'w') {
      return this.player1TimeConsumed + (new Date(Date.now()).getTime() - this.lastMoveTime.getTime());
    }
    return this.player1TimeConsumed;
  }

  getPlayer2TimeConsumed() {
    if (this.board.turn() === 'b') {
      return this.player2TimeConsumed + (new Date(Date.now()).getTime() - this.lastMoveTime.getTime());
    }
    return this.player2TimeConsumed;
  }

  async resetAbandonTimer() {
    if (this.timer) {
      clearTimeout(this.timer);
    }
    this.timer = setTimeout(() => {
      this.endGame("ABANDONED", this.board.turn() === 'b' ? 'WHITE_WINS' : 'BLACK_WINS');
    }, 60 * 1000);
  }

  async resetMoveTimer() {
    if (this.moveTimer) {
      clearTimeout(this.moveTimer)
    }
    const turn = this.board.turn();
    const timeLeft = GAME_TIME_MS - (turn === 'w' ? this.player1TimeConsumed : this.player2TimeConsumed);

    this.moveTimer = setTimeout(() => {
      this.endGame("TIME_UP", turn === 'b' ? 'WHITE_WINS' : 'BLACK_WINS');
    }, timeLeft);
  }

  async endGame(status: GAME_STATUS, result: GAME_RESULT) {

    let gameDb = await db.game.findFirst({
      where :{
        id : this.gameId
      }
    })
    if (!gameDb){
      return
    }
    if (gameDb?.status == "IN_PROGRESS"){
    let  updatedGame = await db.game.update({
      data: {
        status,
        result: result,
      },
      where: {
        id: this.gameId,
      },
      include: {
        moves: {
          orderBy: {
            moveNumber: 'asc',
          },
        },
        blackPlayer: true,
        whitePlayer: true,
      }
    });
    await updatePlayersRating(gameDb)
    SocketManager.getInstance().broadcast(
      this.gameId,
      JSON.stringify({
        type: GAME_ENDED,
        payload: {
          result,
          status,
          moves: updatedGame.moves,
          blackPlayer: {
            id: updatedGame.blackPlayer.id,
            name: updatedGame.blackPlayer.name,
          },
          whitePlayer: {
            id: updatedGame.whitePlayer.id,
            name: updatedGame.whitePlayer.name,
          },
        },
      }),
    );
    // clear timers
    this.clearTimer();
    this.clearMoveTimer();
  }

  clearMoveTimer() {
    if(this.moveTimer) clearTimeout(this.moveTimer);
  }
}
  setTimer(timer: NodeJS.Timeout) {
    this.timer = timer;
  }

  clearTimer() {
    if (this.timer) clearTimeout(this.timer);
  }
}

File: apps/ws-rust/src/game/game.rs
use crate::db::models::{Game, GameResult, GameStatus, Move, User};
use crate::messages::{GameEndedPayload, MovePayload};
use crate::socket_manager::SocketManager;
use chrono::{DateTime, Duration, Utc};
use shakmaty::{Chess, Position};
use std::collections::HashMap;
use std::sync::Arc;
use tokio::sync::{Mutex, RwLock};
use tokio::time::{sleep, timeout};

const GAME_TIME_MS: u64 = 10 * 60 * 60 * 1000; // 10 hours maybe we can change it later

pub struct GameState {
    game: Arc<RwLock<Game>>,
    position: Position,
    chess: Chess,
    player1_time_consumed: Arc<Mutex<u64>>,
    player2_time_consumed: Arc<Mutex<u64>>,
    start_time: DateTime<Utc>,
    last_move_time: Arc<Mutex<DateTime<Utc>>>,
    abandon_timer: Arc<Mutex<Option<tokio::task::JoinHandle<()>>>>,
    move_timer: Arc<Mutex<Option<tokio::task::JoinHandle<()>>>>,
}

impl GameState {
    pub async fn new(game: Game) -> Self {
        let mut chess = Chess::default();
        let position = chess.position();
        let start_time = Utc::now();
        let last_move_time = Arc::new(Mutex::new(start_time));

        GameState {
            game: Arc::new(RwLock::new(game)),
            position,
            chess,
            player1_time_consumed: Arc::new(Mutex::new(0)),
            player2_time_consumed: Arc::new(Mutex::new(0)),
            start_time,
            last_move_time: last_move_time.clone(),
            abandon_timer: Arc::new(Mutex::new(None)),
            move_timer: Arc::new(Mutex::new(None)),
        }
    }

    pub async fn make_move(&mut self, move_payload: MovePayload, user_id: &str) -> Result<(), String> {
        let game = self.game.read().await;

        if (self.chess.turn() == shakmaty::Color::White
            && game.white_player.id != user_id)
            || (self.chess.turn() == shakmaty::Color::Black
                && game.black_player.as_ref().unwrap().id != user_id)
        {
            return Err("Not your turn".to_string());
        }

        if game.result.is_some() {
            return Err("Game is already completed".to_string());
        }

        let move_timestamp = Utc::now();
        let move_str = format!("{}{}", move_payload.from, move_payload.to);

        let move_obj = match self.chess.move_str(&move_str) {
            Ok(m) => m,
            Err(e) => return Err(format!("Invalid move: {}", e)),
        };

        self.chess.apply_move(move_obj);
        self.position = self.chess.position();

        let time_taken = move_timestamp.timestamp_millis() - self.last_move_time.lock().await.timestamp_millis();

        if self.chess.turn() == shakmaty::Color::Black {
            *self.player1_time_consumed.lock().await += time_taken as u64;
        } else {
            *self.player2_time_consumed.lock().await += time_taken as u64;
        }

        let move_data = Move {
            id: uuid::Uuid::new_v4().to_string(),
            game_id: game.id.clone(),
            move_number: game.moves.len() as u64 + 1,
            from: move_obj.from.to_string(),
            to: move_obj.to.to_string(),
            comments: None,
            start_fen: self.chess.fen().to_string(),
            end_fen: self.position.fen().to_string(),
            time_taken: Some(time_taken as u64),
            created_at: move_timestamp,
        };

        let mut game = self.game.write().await;
        game.moves.push(move_data);
        game.current_fen = self.position.fen().to_string();
        *self.last_move_time.lock().await = move_timestamp;

        self.reset_abandon_timer().await;
        self.reset_move_timer().await;

        SocketManager::broadcast(
            game.id.clone(),
            messages::Message {
                r#type: messages::MessageType::Move,
                payload: messages::MovePayload {
                    move_payload,
                    player1_time_consumed: *self.player1_time_consumed.lock().await,
                    player2_time_consumed: *self.player2_time_consumed.lock().await,
                },
            },
        ).await;

        if self.chess.is_game_over() {
            let result = if self.chess.is_draw() {
                GameResult::Draw
            } else if self.chess.turn() == shakmaty::Color::Black {
                GameResult::WhiteWins
            } else {
                GameResult::BlackWins
            };

            self.end_game(GameStatus::Completed, result).await;
        }

        Ok(())
    }

    async fn reset_abandon_timer(&mut self) {
        let game_id = self.game.read().await.id.clone();
        let abandon_timer = tokio::spawn(async move {
            sleep(Duration::from_secs(60)).await;
            let result = if Chess::default().turn() == shakmaty::Color::Black {
                GameResult::WhiteWins
            } else {
                GameResult::BlackWins
            };
            SocketManager::broadcast(
                game_id,
                messages::Message {
                    r#type: messages::MessageType::GameEnded,
                    payload: GameEndedPayload {
                        result,
                        status: GameStatus::Abandoned,
                    },
                },
            ).await;
        });

        let mut abandon_timer_lock = self.abandon_timer.lock().await;
        *abandon_timer_lock = Some(abandon_timer);
    }

    async fn reset_move_timer(&mut self) {
        let game_id = self.game.read().await.id.clone();
        let turn = self.chess.turn();
        let time_left = GAME_TIME_MS
            - if turn == shakmaty::Color::White {
                *self.player1_time_consumed.lock().await
            } else {
                *self.player2_time_consumed.lock().await
            };

        let move_timer = tokio::spawn(async move {
            let result = timeout(Duration::from_millis(time_left), async {
                let result = if turn == shakmaty::Color::Black {
                    GameResult::WhiteWins
                } else {
                    GameResult::BlackWins
                };
                SocketManager::broadcast(
                    game_id,
                    messages::Message {
                        r#type: messages::MessageType::GameEnded,
                        payload: GameEndedPayload {
                            result,
                            status: GameStatus::TimeUp,
                        },
                    },
                ).await;
            }).await;
        });

        let mut move_timer_lock = self.move_timer.lock().await;
        *move_timer_lock = Some(move_timer);
    }

    async fn end_game(&mut self, status: GameStatus, result: GameResult) {
        let mut game = self.game.write().await;
        game.status = status;
        game.result = Some(result);

        SocketManager::broadcast(
            game.id.clone(),
            messages::Message {
                r#type: messages::MessageType::GameEnded,
                payload: GameEndedPayload { result, status },
            },
        ).await;
   }

}